# Exploring Hackers News Posts

In this project, we'll compare two different types of posts from [Hacker News](https://news.ycombinator.com/), a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either `Ask HN` or `Show HN`.

Users submit `Ask HN` posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit `Show HN` posts to show the Hacker News community a project, product, or just generally something interestin

We'll specifically compare these two types of posts to determine the following:

- Do `Ask HN` or `Show HN` receive more comments on average    
- Do posts created at a certain time receive more comments on average?
    
It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.ubmissions.ons.g.

# Introduction

First, we'll read the data and remove the headers.


In [41]:
import csv

with open('./dataset/hacker_news.csv') as data:
    hn = list(csv.reader(data)) # data as list of lists

hn_header = hn[0]
hn = hn[1:]  # data without headers

print(f'\nHEADER:\n\n{hn_header}')
print(f'\nFIRST 3 ROWS:\n\n{hn[:3]}')



HEADER:

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']

FIRST 3 ROWS:

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']]


# Extracting Ask HN ans Show HN Posts

First, we'll identify posts that begin with either `Ask HN` or `Show HN` and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [74]:
ask_posts = []
show_posts = []
other_posts = []

# With this method I have to put the exact title, so it is not valid

#for row in hn:
#    post = row[1]
#    if post == 'Ask HN':
#        ask_posts.append(row)
#    elif post == 'Show HN':
#        show_posts.append(row)
#
#print(ask_posts[:3])
#print(show_posts[:3])

for row in hn:
    post = row[1]
    if post.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif post.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)

print(ask_posts[:3])
#print(show_posts[:3])
#print(other_posts[:3])

print(f'\nAsk HN posst: {len(ask_posts)} \nShow HN posts: {len(show_posts)} \nOther posts: {len(other_posts)}')

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14']]

Ask HN posst: 1744 
Show HN posts: 1162 
Other posts: 17194


# Average Number of Comments for Ask HN and Show HN Posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [58]:

ask_comments = 0
show_comments = 0
other_comments = 0

# Average of comments in 'Ask HN' posts
for post in ask_posts:
    n_comments = int(post[4])
    ask_comments += n_comments

# Average of comments in 'Show HN' posts
for post in show_posts:
    n_comments = int(post[4])
    show_comments += n_comments

avg_ask_comments = ask_comments / len(ask_posts)
avg_show_comments = show_comments / len(show_posts)

print(f'Ask HN posts have a total of {ask_comments} comments, with an average of {round(avg_ask_comments,2)} per post.')
print(f'Show HN posts have a total of {show_comments} comments, with an average of {round(avg_show_comments,2)} per post.')


Ask HN posts have a total of 24483 comments, with an average of 14.04 per post.
Show HN posts have a total of 11988 comments, with an average of 10.32 per post.


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.

# Number of Ask posts and Comments by Hour created 

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [90]:
import datetime as dt

date_format = "%m/%d/%Y %H:%M"
comments_hour = {}
post_hour = {}

for row in ask_posts:
    creation_date = row[-1]
    comments = int(row[4])
    time = dt.datetime.strptime(creation_date,date_format).strftime('%H')
    if time not in comments_hour:
        comments_hour[time] = comments
        post_hour[time] = 1
    elif time in comments_hour:
        comments_hour[time] += comments
        post_hour[time] += 1

print(f'Posts created by hour:\n {post_hour}')
print(f'\nComments created by hour:\n {comments_hour}')


Posts created by hour:
 {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}

Comments created by hour:
 {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


# Average Number of Comments for Ask HN Posts by Hour



In [93]:
avg_hour = []

for time in comments_hour:
    avg_comments = comments_hour[time] / post_hour[time]
    avg_hour.append([time, avg_comments])

avg_hour  

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

# Sorting and Printing Values from a List of Lists

Although we now have the results we need, this format makes it difficult to identify the hours with the highest values. Let's finish by sorting the list of lists and printing the five highest values in a format that's easier to read.

In [97]:
swap_avg_hour = []

for row in avg_hour:
    swap_avg_hour.append([row[1], row[0]])

sorted_swap = sorted(swap_avg_hour, reverse=True)

sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [100]:

print("Top 5 Hours for 'Ask HN' Comments")
for avg, time in sorted_swap[:5]:
    print(
        f"{dt.datetime.strptime(time, '%H').strftime('%H:%M')}: {avg:.2f} average comments per post"
    )

Top 5 Hours for 'Ask HN' Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/datasets/hacker-news/hacker-news-posts), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.